saw-central/src/SAWCentral/Crucible/Common/Vacuity.hs

```diff
@@ -0,0 +1,108 @@
+import SAWCentral.Proof
+import SAWCentral.TopLevel
+import SAWCentral.Value
+import SAWCentral.Options
+import qualified SAWCentral.Crucible.Common as Common
+import           SAWCentral.Crucible.Common.MethodSpec (CrucibleMethodSpecIR)
+import qualified SAWCentral.Crucible.Common.MethodSpec as MS
+
+type AssumptionReason = (MS.ConditionMetadata, String)
+
+-- | Minimal backend-independent context
+data GenericCrucibleContext sym = GenericCrucibleContext
+  { gccSym :: sym  -- What4 symbolic backend
+  }
+
+-- | Top-level check and minimal core reporter
+checkAssumptionsForContradictions ::
+  GenericCrucibleContext Common.Sym ->
+  CrucibleMethodSpecIR ext ->
+  ProofScript () ->
+  [Crucible.LabeledPred Term AssumptionReason] ->
+  TopLevel ()
+checkAssumptionsForContradictions gcc methodSpec tactic assumptions =
+  whenM
+    (assumptionsContainContradiction gcc methodSpec tactic assumptions)
+    (computeMinimalContradictingCore gcc methodSpec tactic assumptions)
+
+-- | Check if the assumptions are contradictory
+assumptionsContainContradiction ::
+  GenericCrucibleContext Common.Sym ->
+  CrucibleMethodSpecIR ext ->
+  ProofScript () ->
+  [Crucible.LabeledPred Term AssumptionReason] ->
+  TopLevel Bool
+assumptionsContainContradiction gcc methodSpec tactic assumptions = do
+  let sym = gccSym gcc
+  st <- io $ Common.sawCoreState sym
+  let sc = saw_ctx st
+  let ploc = methodSpec ^. MS.csLoc
+
+  (goal', pgl) <- io $
+   do
+    -- conjunction of all assumptions
+    assume <- scAndList sc (toListOf (folded . Crucible.labeledPred) assumptions)
+    -- implies falsehood
+    goal   <- scImplies sc assume =<< toSC sym st (W4.falsePred sym)
+    goal'  <- boolToProp sc [] goal
+    return (goal',
+            ProofGoal
+              { goalNum     = 0
+              , goalType    = "vacuousness check"
+              , goalName    = "vacuousness_check"
+              , goalLoc     = show (W4.plSourceLoc ploc) ++ " in " ++ show (W4.plFunction ploc)
+              , goalDesc    = "vacuousness check"
+              , goalSequent = propToSequent goal'
+              , goalTags    = mempty
+              })
+
+  res <- runProofScript tactic goal' pgl Nothing
+           "vacuousness check" False False
+  case res of
+    ValidProof _ _     -> return True
+    InvalidProof _ _ _ -> return False
+    UnfinishedProof _  ->
+      -- TODO? is this the right behavior?
+      do printOutLnTop Warn "Could not determine if preconditions are vacuous"
+         return True
+
+-- | Try to find the smallest contradictory subset
+computeMinimalContradictingCore ::
+  GenericCrucibleContext Common.Sym ->
+  CrucibleMethodSpecIR ext ->
+  ProofScript () ->
+  [Crucible.LabeledPred Term AssumptionReason] ->
+  TopLevel ()
+computeMinimalContradictingCore gcc methodSpec tactic assumes = do
+  printOutLnTop Warn "Contradiction detected! Computing minimal core of contradictory assumptions:"
+  let cores = sortBy (compare `on` length) (subsequences assumes)
```

Comment:
This is missing a comment from the original code in `SAWCrucible.Crucible.Builtins.LLVM`, which would be nice to include:

```suggestion
  -- test subsets of assumptions of increasing sizes until we find a
  -- contradictory one
  let cores = sortBy (compare `on` length) (subsequences assumes)
```